### PR TITLE
Static Cache: no mandatory `cache_positions` input

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Removes the hard requirement of `cache_position` from the public model classes (e.g. classes that can be loaded with `AutoModel`). Its contents can be derived from the cache's `get_seq_length()`.

________________________________________________________
### Performance and Quality checks

Slow tests run, getting the same outcome as in `main`:
- [x] `RUN_SLOW=1 py.test tests/models/llama/test_modeling_llama.py -vv`
- [x] `RUN_SLOW=1 py.test tests/models/gemma/test_modeling_gemma.py -vv`
- [x] `RUN_SLOW=1 py.test tests/test_cache_utils.py -vv` [Note: two new tests were added here]

👉 Note that these tests ensure that `torch.compile` works, with and without `attention_mask` being passed.

Local benchmark (RTX3090, tiny llama) -- no changes
(main)
Screenshot 2024-02-22 at 11 54 26

(this pr)
Screenshot 2024-02-22 at 17 57 13

### CODE REVIEW

**Comment by gante** [comment]
cc @fxmarty :)

File: src/transformers/models/gemma/modeling_gemma.py [hunk 1]
```diff
@@ -864,13 +863,24 @@ def forward(
...
+        # applied on all batch members. Left-padding on all batch members can be detected from the `attention_mask`.
+        cache_position = torch.max(position_ids, dim=0).values
+        if attention_mask is None:
+            padded_offset = 0
+        else:
+            padded_offset = (1 - torch.sum(attention_mask.to(torch.int64), dim=0).clamp(max=1)).cumsum(-1)
+            padded_offset = torch.cat(
+                (torch.zeros((1,), dtype=torch.int64, device=padded_offset.device), padded_offset)
+            )[-cache_position.shape[0] - 1 : -1]
+        cache_position = cache_position + padded_offset
```
Comment (ArthurZucker) [inline src/transformers/models/gemma/modeling_gemma.py:883]: you are using sum + clamp + cumsum + cat , costly for : long sequence length, 4D attention mask provided to the model.

**Review by ArthurZucker — COMMENTED** [review]
I don't think that is the way we should go. For now have not seen issue with having a new argument, we added `paddin_mask` at some point.
IMO we should move to having `cache_positions` in generate as this is more explicit, easier to maintain and less error prone. 
More than that, these operation are input dependent while being vectorized they add complexity when there should not be!

**Comment by fxmarty** [comment]
No specific opinon, just that using `cache_position` in the modeling code when `use_cache=False` is not very intuitive to me: https://github.com/huggingface/transformers/blob/75ed76eceaf9b20c7ec37395e4f5d491135186f9/src/transformers/models/llama/modeling_llama.py#L975-L983

Only thing I think: `cache_position` (if it remains) should be a strictly optional argument & the model should work without for BC.

File: src/transformers/models/llama/modeling_llama.py [hunk 2]
```diff
@@ -972,15 +971,24 @@ def forward(
...
+        # applied on all batch members. Left-padding on all batch members can be detected from the `attention_mask`.
+        cache_position = torch.max(position_ids, dim=0).values
+        if attention_mask is None:
+            padded_offset = 0
+        else:
+            padded_offset = (1 - torch.sum(attention_mask.to(torch.int64), dim=0).clamp(max=1)).cumsum(-1)
+            padded_offset = torch.cat(
+                (torch.zeros((1,), dtype=torch.int64, device=padded_offset.device), padded_offset)
+            )[-cache_position.shape[0] - 1 : -1]
+        cache_position = cache_position + padded_offset
```
Comment (fxmarty) [inline src/transformers/models/llama/modeling_llama.py:991]: @gante was the prior 

```python
cache_position = torch.arange(past_seen_tokens, past_seen_tokens + inputs_embeds.shape[1], device=inputs_embeds.device)
```

wrong?

**Comment by gante** [comment]
> No specific opinon, just that using cache_position in the modeling code when use_cache=False is not very intuitive to me:

@fxmarty I agree here, that it should be renamed :) e.g. `data_position` would be more appropriate, as it may be used without cache. Precise variable names help us in the long run.

> IMO we should move to having cache_positions in generate as this is more explicit, easier to maintain and less error prone.
More than that, these operation are input dependent while being vectorized they add complexity when there should not be!

@ArthurZucker I strongly disagree, more redundant inputs result in more bugs 🐛  From the full `input_ids` and `pad_token_id` we can derive `attention_mask`, `position_ids`, and `cache_positions` -- it's not like we are giving more options to the users, there is only one set of valid tensors given the first two. In other words, we give users three different inputs to get things wrong! As a result, we have to add safety warnings and exceptions, such as:
1. [warning if there is a pad token in the input but no attention mask](https://github.com/huggingface/transformers/blob/c8d98405a8f7b0e5d07391b671dcc61bb9d7bad5/src/transformers/models/gpt2/modeling_gpt2.py#L775C18-L775C55)
2. [raising an exception if cache positions are not passed](https://github.com/huggingface/transformers/blob/c8d98405a8f7b0e5d07391b671dcc61bb9d7bad5/src/transformers/models/llama/modeling_llama.py#L977)
3. [raise an exception if the attention mask is not of the right type](https://github.com/huggingface/transformers/blob/c8d98405a8f7b0e5d07391b671dcc61bb9d7bad5/src/transformers/models/gpt2/modeling_gpt2.py#L261)
4. [warning if both attention mask and pad token are missing in generate](https://github.com/huggingface/transformers/blob/c8d98405a8f7b0e5d07391b671dcc61bb9d7bad5/src/transformers/generation/utils.py#L1360)
(plus other error-prone combinations that we are not even checking 😬)

This is arguably much more work than enforcing the correct pairing in the models themselves! On top of that, other functions like `generate` or `pipeline` have to create the correct logic anyways (e.g. [create attn mask](https://github.com/huggingface/transformers/blob/c8d98405a8f7b0e5d07391b671dcc61bb9d7bad5/src/transformers/generation/utils.py#L469); [create position ids](https://github.com/huggingface/transformers/blob/c8d98405a8f7b0e5d07391b671dcc61bb9d7bad5/src/transformers/models/gpt2/modeling_gpt2.py#L1018)), so why not push the logic to the models and prevent these bugs from occurring everywhere? 🤗 

My goal with this PR is to iron out future sources of bugs before we roll out these changes to other models. Better than well-explained interfaces is... no need for that interface at all!

In terms of complexity, it will always exist. The difference is whether it exists outside `generate` (=users can get it wrong, one more argument in `generate`), inside `generate` (=users can get custom generation loops wrong, one more argument in `forward`), or inside `forward` (=we have to maintain the logic, but it is easy to do it right without bugs). In my experience, if some input can be used incorrectly, it WILL be used incorrectly. [In the same direction, I think we should get rid of `attention_mask` and `position_ids`, but that would be a breaking change 😉 ]

In terms of performance, we can see that it is negligible in eager forwards and none in static forwards. Again, if not implemented here, it will be implemented in `generate`, resulting in the same net performance.

File: src/transformers/models/gemma/modeling_gemma.py (hunk 1 above)
Comment (gante) [reply src/transformers/models/gemma/modeling_gemma.py:883]: Not quite, see performance numbers above :)

File: src/transformers/models/llama/modeling_llama.py (hunk 2 above)
Comment (gante) [reply src/transformers/models/llama/modeling_llama.py:991]: No, but it was not compatible with static cache -- users had to pass `position_ids`, `cache_positions`, and `attention_mask` to get the right output, and these are redundant tensors. 

My goal with this PR was to remove the need to pass `cache_positions` :) (full rationale [here](https://github.com/huggingface/transformers/pull/29221#issuecomment-1961626744))

File: src/transformers/cache_utils.py [hunk 3]
```diff
@@ -398,16 +398,9 @@ def update(
...
-        )
-
-    def get_usable_length(self, new_sequence_length=None, layer_idx: Optional[int] = 0) -> int:
-        # TODO: Fix once the stateful `int` bug in PyTorch is fixed.
-        raise ValueError(
-            "get_seq_length is not implemented for StaticCache. Please refer to https://github.com/huggingface/transformers/pull/29114."
-        )
+        # Occupied cache == any slot in the 3rd dim (sequence length) holds a non-zero value. To save on compute, let's
+        # limit the check to the first batch member and head dimension.
+        return (self.key_cache[0, 0].any(dim=-1)).sum()
```
[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/cache_utils.py:403]: @fxmarty yes, this version works! e.g. try running `RUN_SLOW=1 py.test tests/test_cache_utils.py::CacheIntegrationTest::test_static_cache_greedy_decoding_pad_left_0_eager`, which confirms that all dynamic, eager static cache, and compiled static cache get the same sensible outputs.

> I'm not a fan of this. This would be run at each forward. Couldn't we just increment somewhere?

I am very much aligned with this! Indeed that was my original plan by adding `seen_tokens` to the original cache implementation. However, as you mentioned, we haven't found a way to make it work at compile time. Perhaps we shall add a TODO to revisit this in the future, after the issue you opened on the PyTorch side has been addressed?

**Review by ArthurZucker — COMMENTED** [review]
Thanks for your great explanation.  
TLDR; I think part of this choice comes down to being `implicit` vs `explicit`.
**Your approach works**, but it's implicit. Both have ➕ and ➖ for maintainability. 
I am leaning a lot more towards the `explicit` for a few reasons:

1. If you generate the `cache_positions` in the model with `torch.arange`, the generations you will get with cuda graphs will be wrong precisely because the cache positions are not given as input. I tried this and it does not work. 
2. ➕ Instead of relying on hidden `seen_tokens` and 3 different `get_length`, `get_usable_lenght`, `get_max_length`, you just need `cache_positions` in generate.
3. ➕ Stateless is less error prone, `cache_positions` is also pretty known from GPT-Fast.
4. > From the full input_ids and pad_token_id we can derive attention_mask, position_ids, and cache_positions
Is not true: if you want to do paged attention / packed training with a custom 4D attention mask or anything a bit outside the classic forward (any custom Cache class), then you cannot only rely on input_ids and pad token.
5. The `generate` function **should** keep track of where it is, for simplicity but also because otherwise your not really able to manipulate anything outside the forward of the model. So at the end of the day you have a loop in generate, which is the only `public` place where we use the past_key_values, but you don't let it handle the `cache_positions` which is a bit strange! 

Let's keep the cache positions, gradually remove calls to `get_seq_length`, and explicitly pass the arguments to the model instead of relying on the model.

File: src/transformers/models/gemma/modeling_gemma.py [hunk 4]
```diff
@@ -1101,14 +1110,24 @@ def forward(
...
         past_length = 0
         if past_key_values is not None:
             if isinstance(past_key_values, Cache):
-                cache_length = past_key_values.get_seq_length()
-                past_length = past_key_values.seen_tokens
+                past_length = (
+                    cache_position[-1] + 1 if cache_position is not None else past_key_values.get_seq_length()
+                )
                 max_cache_length = past_key_values.get_max_length()
+                cache_length = past_length if max_cache_length is None else min(max_cache_length, int(past_length))
```
Comment (gante) [inline src/transformers/models/gemma/modeling_gemma.py:1129]: This restructure prioritizes `cache_position`, falling back to `.get_seq_length()` in its absence. This replaces `seen_tokens`, which is now deprecated.

Note that `past_length` [all seen tokens] and `cache_length` [tokens in the cache] are both needed, otherwise `SinkCache` won't work.

File: src/transformers/models/gemma/modeling_gemma.py [hunk 5]
```diff
@@ -1141,19 +1160,11 @@ def prepare_inputs_for_generation(
...
-                past_length = cache_position[-1] + 1
-            input_ids = input_ids[:, past_length:]
-            position_ids = position_ids[:, past_length:]
-
         # TODO @gante we should only keep a `cache_position` in generate, and do +=1.
         # same goes for position ids. Could also help with continued generation.
-        cache_position = torch.arange(past_length, past_length + position_ids.shape[-1], device=position_ids.device)
+        input_length = position_ids.shape[-1] if position_ids is not None else input_ids.shape[-1]
+        cache_position = torch.arange(past_length, past_length + input_length, device=input_ids.device)
+        position_ids = position_ids.contiguous() if position_ids is not None else None
```
Comment (gante) [inline src/transformers/models/gemma/modeling_gemma.py:1167]: This is already on main (see [here](https://github.com/huggingface/transformers/blob/2209b7afa04b3a6366350065f541e9248d6663c2/src/transformers/models/gemma/modeling_gemma.py#L1176)), not sure why this shows up 👀

**Comment by ArthurZucker** [comment]
@gante could you rebase / fix meges and I ll review!

File: src/transformers/models/gemma/modeling_gemma.py [hunk 6]
```diff
@@ -859,14 +863,19 @@ def forward(
...
+                past_seen_tokens = static_cache.get_seq_length()
+            else:
+                if not isinstance(past_key_values, Cache):
+                    past_key_values = DynamicCache.from_legacy_cache(past_key_values)
+                past_seen_tokens = past_key_values.get_seq_length()
 
         if cache_position is None:
-            cache_position = torch.arange(
-                past_seen_tokens, past_seen_tokens + inputs_embeds.shape[1], device=inputs_embeds.device
+            # `torch.compile`-friendly `torch.arange` from a shape
```
Comment (ArthurZucker) [inline src/transformers/models/gemma/modeling_gemma.py:876]: does that also fix the ONNX export we had?

**Review by ArthurZucker — COMMENTED** [review]
Late review, not sure what is the status. 
Overall the goal is to not have to pass cache positions, but the logic is very complicated, while asking users to pass the cache positions seems a lot simpler no?

**Comment by gante** [comment]
Closing this PR and other cache PRs, as we want to move in the opposite direction (static cache behaving like the other caches)
